### ISSUE TITLE
optimization.splitChunks.maxSize throws error or causes build to hang

### ISSUE THREAD

**Issue by chrisckc** [webpack/webpack#8407]
# Bug report

**What is the current behavior?**

An error is thrown when attempting to use optimization.splitChunks.maxSize or optimization.splitChunks.cacheGroups.{cachegroupname}.maxSize

Example config:
```
optimization: {
    splitChunks: {
      chunks: "initial",
      //maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          //enforce: true, // create chunk regardless of the size of the chunk
          maxSize: 10000
        }
      }
    }
  },
```

throws the following error:
```
setup/node_modules/webpack/lib/util/deterministicGrouping.js:168
				leftSize += group.nodes[left].size;
				                              ^
TypeError: Cannot read property 'size' of undefined
    at module.exports (......../minimal-react-webpack-babel-setup/node_modules/webpack/lib/util/deterministicGrouping.js:168:35)
    at compilation.hooks.optimizeChunksAdvanced.tap.chunks (......../minimal-react-webpack-babel-setup/node_modules/webpack/lib/optimize/SplitChunksPlugin.js:797:23)
```

Setting maxSize at the splitChunks level causes the same error.

Uncommenting the line `//enforce: true,` removes the error but causes the build to hang.

**If the current behavior is a bug, please provide the steps to reproduce.**

Clone the following repo:
https://github.com/chrisckc/minimal-react-webpack-babel-setup

Follow the instructions in the readme

This error was first spotted in a more complex app, so i found and forked a less complex app repo to verify the bug.

I can't make the demo repo any simpler because there needs to be enough node modules available to recreate the issue.

**What is the expected behavior?**

It should create multiple 'vendors' chunks each capped at maxSize or as close as possible.

**Other relevant information:**
webpack version:  v4.23.1
Node.js version: v8.11.4
Operating System: MacOS 10.13.4
Additional tools: VSCode, mac terminal

**Comment by chrisckc** [comment]
Looks the same as this issue:
https://github.com/webpack/webpack/issues/7455

which is claimed to be fixed by this:
https://github.com/webpack/webpack/pull/7651

**Comment by chrisckc** [comment]
@sokra Just had a look at the change which sets max size using `Math.max(minSize, maxSize)`

This prevents the crash but the hang during build remains when i use either `enforce: true` or a minSize less then maxSize such as `minSize: 1000,` in the config as follows:

```
optimization: {
    splitChunks: {
      chunks: "initial",
      minSize: 1000,
      //maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          //enforce: true, // create chunk regardless of the size of the chunk
          maxSize: 10000
        }
      }
    }
  },
```

The issue seems to be caused by a loop as the node process starts to use all of the CPU and consumes the RAM at a steady rate.

**Comment by chrisckc** [comment]
If maxSize is specified at the the splitChunks level as per the following config, it works fine:

```
optimization: {
    splitChunks: {
      chunks: "initial",
      minSize: 1000,
      maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          //enforce: true, // create chunk regardless of the size of the chunk
          //maxSize: 10000
        }
      }
    }
  },
```

The output is as follows:

```
bundle.js  8.62 KiB       0  [emitted]  main~26a0f563
vendors~0928ebd2.11051f8abac033be4592.chunk.js  96.1 KiB       1  [emitted]  vendors~0928ebd2
vendors~53127b40.63e088532ae1ece53785.chunk.js  4.76 KiB       2  [emitted]  vendors~53127b40
vendors~c1dd23ef.781dbcd770ccb43df9c7.chunk.js  7.29 KiB       3  [emitted]  vendors~c1dd23ef
```
The larger chunk is due to the single large 'react-dom' module which is 96KB

**Comment by chrisckc** [comment]
There is also another issue when maxSize is used, large single modules larger than maxSize seem to be processed as they are split into 2 chunks, however the second chunk is always almost empty containing no module code.

I don't know for sure but i am guessing that it is not possible to split a single large module into smaller parts?

Either way this is a bug because if the module cant be split it should be left alone and remain as a single chunk, if it can be split then it should be split evenly.

using the following config which splits out 'react-dom'

```
optimization: {
    splitChunks: {
      chunks: "initial",
      minSize: 1000,
      maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        reactDom: {
          test: /[\\/]node_modules[\\/]react-dom[\\/]/,
          name: "vendor.react-dom",
          enforce: true,
          priority: 20
        },
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          //enforce: true, // create chunk regardless of the size of the chunk
          //maxSize: 10000
        }
      }
    }
  },
```

results in the following chunks:

```
bundle.js   8.63 KiB       0  [emitted]  main~26a0f563
vendor.react-dom~0928ebd2.18d3479e1d1e127ecfc1.chunk.js   96.1 KiB       1  [emitted]  vendor.react-dom~0928ebd2
vendor.react-dom~6a83488c.21c9b1ca71c8af4e38ae.chunk.js  312 bytes       2  [emitted]  vendor.react-dom~6a83488c
         vendors~0ba66fd7.70e5f1155bc6aa83b645.chunk.js   1.03 KiB       3  [emitted]  vendors~0ba66fd7
         vendors~53127b40.130f3482b0cc0e62c6ca.chunk.js   4.76 KiB       4  [emitted]  vendors~53127b40
         vendors~55119a63.5f817c5f775bf8ddc54f.chunk.js   6.08 KiB       5  [emitted]  vendors~55119a63
```
Notice the tiny 'react-dom' chunk in the output.

This is how i spotted the original issue as i tried to workaround that behaviour by using maxSize at the cacheGroup level instead of the splitChunks level.

**Comment by chrisckc** [comment]
@sokra  I can confirm that https://github.com/webpack/webpack/pull/8416
partially fixes the hang issue.

It fixes the following config whre maxSize is specified at the cacheGroup level:

```
optimization: {
    splitChunks: {
      chunks: "initial",
      minSize: 1000,
      //maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          //enforce: true, // create chunk regardless of the size of the chunk
          maxSize: 10000
        }
      }
    }
  },
```

but does not fix the hang when enforce: true is used in the cacheGroup


```
optimization: {
    splitChunks: {
      chunks: "initial",
      minSize: 1000,
      //maxSize: 10000,
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          enforce: true, // create chunk regardless of the size of the chunk
          maxSize: 10000
        }
      }
    }
  },
```

**Comment by chrisckc** [comment]
Also hangs when using 'minSize: 0' instead of 'enforce: true' at cacheGroup level

**Comment by sokra** [comment]
hmm... I can't reproduce it with your repo. Could you update it to the problematic configuration?

**Comment by sokra** [comment]
That is working as expected in my opinion. These chunks are not empty, they contain only a small module. As `minSize` is 0 (`enforce: true` = `minSize: 0`) this is valid. The algorithm chooses chunks by heuristic within minSize and maxSize bounds. Here it chooses to make smaller chunks. It assumes that `./node_modules/scheduler/index.js` (`vendors~c171cc99.31b3b569dd046400b75b.chunk.js`) and `./node_modules/react/index.js` (`vendors~cb5fc66c.bc349775f11c7f447ed9.chunk.js`) are independent from each other. That's not too wrong as they are different npm packages... All the algorithm knows are the filenames. If you want bigger chunks, use a higher `minSize` value. The default is 30kb, but even lower values make sense when using HTTP/2.

Try to avoid `enforce: true` as it removes the default restrictions that i. e. prevent too small chunks.

You don't need to `cacheGroups` for npm packages. maxSize should automatically group them together as needed.

If single modules are larger than maxSize they are put alone into a chunk. i. e. this could happen for react-dom or other prebundled libraries. Actually it's a bit bad that they are prebundled, because we are unable to split them and can't fulfill `maxSize` here.

**Comment by chrisckc** [comment]
Thanks for the info, it makes sense now, after some more testing i can see that using MaxSize with 'enforce: true' is not a good combination. As you have pointed out , in the case of react, it splits out the node modules index.js from the rest of the prebundled code.

I did a bit more testing after adding webpack-bundle-analyzer.
In the case of this setup, making sure minSize is greater then 1500 prevents the index split from happening (the stat size of the index is just under that size).

In reality, for HTTP/2 , there is little point using a minSize of than 5000, as i understand it, there is a small size overhead involved in bundle splitting which will become noticeable if splitting into too many chunks.

I can get the same splits (in terms of size) using the built in cacheGroups for this simple example, but i prefer to specify cacheGroups to get more control over the content and naming of the chunks to aid long term cacheability. (combined with optimization.runtimeChunk: true and optimization.moduleIds: 'hashed')

**Comment by chrisckc** [comment]
On the subject of the 'enforce' option, the docs state the following:

> Tells webpack to ignore splitChunks.minSize, splitChunks.maxSize, splitChunks.minChunks, splitChunks.maxAsyncRequests and splitChunks.maxInitialRequests options and always create chunks for this cache group.

I don't think that properly describes the behaviour of the option, it does not state that it also results in minSize being set to zero. The words "Tells webpack to ignore splitChunks.minSize" is a bit misleading.

As i have not really looked at the source, until you explained the behaviour i did not understand why a small shard of a chunk was being split from a module bigger than maxSize.

So in light of the above, how could a cacheGroup be specified such that it is always created but still respects the minSize setting specified at splitChunks level? In other words, always create the chunk but don't break tiny bits from it if it bigger than maxSize?

Would i need to specify each option, maxSize, minChunks, maxAsyncRequests and maxInitialRequests inside each cacheGroup to override any of those options set at splitChunks level?

**Comment by chrisckc** [comment]
It seems i can get around the fact that enforce: true sets the cacheGroup's minSize to zero by adding 'maxSize: Infinity' but it means i then lose out on the option to split the chunk if it is splittable, so it seems the only option is to specify each option individually.

**Comment by chrisckc** [comment]
Thinking about this a little more, would it not be better to have 'enforce: true' behave as follows:

"Always create the chunk regardless of splitChunks.minSize, splitChunks.maxAsyncRequests and splitChunks.maxInitialRequests. When creating the chunk, respect maxSize by attempting to split if bigger then maxSize, but only if this does not violate minSize."

**Comment by sokra** [comment]
Yeah it probably makes sense that enforce doesn't modify the minSize used for maxSize, but changing that would be a breaking change for v4. I can change it for v5.

Anyway try to avoid using enforce.

**Comment by sokra** [comment]
Actually the doc is a bit wrong about ignoring minSize. It's doesn't ignore it, it set the default to 0. If you specify enforce and minSize, it will respect minSize.

**Comment by sokra** [comment]
Maybe we can consider the fact that minSize is ignored for maxSize when enforce is specified as big that I can fix it

**Comment by chrisckc** [comment]
@sokra  I agree, here are some testing results:

> If you specify enforce and minSize, it will respect minSize.

Tested that and found that it does indeed respect minSize when its specified on either splitChunks or the CacheGroup.  **However if maxSize is then specified on the cacheGroup, minSize from splitChunks is not respected**, minSize also has to be added to the cacheGroup as well for it to be respected.

This would suggest that maxSize from splitChunks is not being applied to a cacheGroup which has 'enforce: true', but that is not the case. I found that maxSize from splitChunks is respected when 'enforce: true' is used in cacheGroups, the following config demonstrates:

```
splitChunks: {
      chunks: "initial",
      minSize: 5000,
      maxSize: 10000
      cacheGroups: {
        default: false, // disable the built-in groups, default & vendors (vendors is overwritten below)
        reactDom: {
          test: /[\\/]node_modules[\\/]react-dom[\\/]/,
          name: "vendor.react-dom",
          enforce: true,
          priority: 20
        },
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendors",
          priority: 10,
          enforce: true
        }
      }
    }
```

produces the following chunks:

```
bundle.js  8.61 KiB       0  [emitted]  main~26a0f563
vendor.react-dom~ab68c3a7.2bb28aa6d557c3411cea.chunk.js  96.2 KiB       1  [emitted]  vendor.react-dom~ab68c3a7
         vendors~53127b40.63e088532ae1ece53785.chunk.js  4.76 KiB       2  [emitted]  vendors~53127b40
         vendors~c1dd23ef.6d185629eeec372d514b.chunk.js  7.02 KiB       3  [emitted]  vendors~c1dd23ef
```

notice that vendors has correctly been split because it would be bigger than 10k and vendor.react-dom is left intact because splitting it would violate minSize.

Before the fixes referenced here, which i noticed have now been released in v4.26.1, the above config would have created an additional vendor.react-dom chunk that violates minSize.

To sum it up, v4.26.1 solves the original issues, there just remains this edge case when using 'enforce: true' and MaxSize together on a cacheGroup.

The main use case for enforce, in my mind, is to override splitChunks.maxInitialRequests and splitChunks.maxAsyncRequests so that the chunk would always be created. It makes more sense than having to set maxInitialRequests maxAsyncRequests to Infinity on a cacheGroup. For enforce to make sense it also means minSize has to be ignored when deciding to create the chunk, but as you have said, it should still be respected when attempting to split the chunk.

**Comment by sokra** [comment]
Looks good.

I'm not a big fan of enforcing `name`s on cacheGroups since this can also have drawbacks. But in this end it's a trade-off you can use.

**Comment by sokra** [comment]
> Note: The priority needs to be higher than zero so that it is processed before the built-in cacheGroups. 

The default cacheGroups have negative priority, so even priority 0 (default) can be used.

**Comment by sokra** [comment]
Regarding using `name`. There can't be two chunks with the same name. So if you try to give two chunks the same name they will merge. The same will happen with splitChunks. If you pass a name chunk could merge.

Let's compare these configurations: 
`{ test: /node_modules/ }` (1) and 
`{ test: /node_modules/, name: "vendors" }` (2).

We assume a scenario like this: 

``` js
// module a
import "lodash";
import "react";

// module b
import "lodash";

// entry
import("./a"); // A
import("./b"); // B
```

Without splitChunks this will result in the following chunks:

```
chunk "main":
  - entry.js

chunk "a":
  - a.js
  - lodash
  - react

chunk "b":
  - b.js
  - lodash

A loads chunk "a"
B loads chunk "b"
```

With configuration (1):

```
chunk "main":
  - entry.js

chunk "a":
  - a.js

chunk "vendors~a~b":
  - lodash

chunk "vendors~a":
  - react

chunk "b":
  - b.js

A loads chunk "a", "vendors~a~b" and "vendors~a"
B loads chunk "b" and "vendors~a~b"
```
  
With configuration (2):

```
chunk "main":
  - entry.js

chunk "a":
  - a.js

chunk "vendors":
  - lodash
  - react

chunk "b":
  - b.js

A loads chunk "a" and "vendors"
B loads chunk "b" and "vendors"
```

This looks innocent, but notice how B loads more code than needed, so it may cause an unnecessary delay.
